### PR TITLE
[RNMobile] Wrap Block and Inspector control behind isSelected flag

### PR DESCRIPTION
**Pull request by lukewalczak**
## Description

Wrapping Inspector and Block controls behind `isSelected` flag to avoid redundant rerenders in blocks:

As wrote below, in the scope of one block it's not a big deal, but in terms of the whole post it can make the difference.

### Button

before | after
--- | ---
b | a

### List

before | after
--- | ---
List_a | List_b

### Image

before | after
--- | ---
Image_a | Image_b

### Verse

before | after
--- | ---
Verse_a | Verse_b

 

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by gziolo** [comment]
In the web version, `isSelected` check is baked into `InspectorControld` and `BlockControls`. They use the block edit context to read `isSelected`. Is there anything in RN that prevents replicating the same handling? I know that portals and therefore slots work differently but maybe there is a way to along both platforms.

**Comment by lukewalczak** [comment]
Hello @gziolo, I didn't check that solution, but definitely it's worth investigating. I will double check that, thanks for hint!

**Comment by lukewalczak** [comment]
Actually we have the same mechanism in native inspector control equivalent and nested children are not rendered until the block is selected, however the InspectorControls itself has to be rendered and its instance is created for every block and each time it's calling useBlockEditContext

Generally the aim of that PR is to avoid redundant renders. In the scope of one block it's not a big deal, but in terms of the whole post it can make the difference.

before | after
--- | ---
b | a

Are there any contraindications to avoid it in the crossplatform blocks such as `Gallery` or `Heading`?

**Comment by gziolo** [comment]
The block edit context (at least for the web) updates only for the block selected and the one unselected. Everything else remains the same with the help of memoization.

Now, when there are child blocks at each level of nesting, `isSelected` won’t change to `true` until those blocks get explicitly selected. So child blocks would only rerender when they somehow read data from their parent. 

That's the theory that might be no longer true, but it means we should fix it on a higher level. It’s OK to add those checks, but ideally it shouldn’t be necessary at all. Is there any complex logic used to compute children elements of block and inspector controls? I can see it being noticeable in the performance tools.

**Comment by lukewalczak** [comment]
> The block edit context (at least for the web) updates only for the block selected and the one unselected. Everything else remains the same with the help of memoization.

Now, when there are child blocks at each level of nesting, isSelected won’t change to true until those blocks get explicitly selected. So child blocks would only rerender when they somehow read data from their parent.

Basically it's not about changing `isSelected` but rerendering `InspectorControls` component on mobile, when there is an update in component which contains controls (block edit). Unfortunately sometimes even rendering the whole `FlatList` makes updates in each block component :|

**Comment by gziolo** [comment]
> Basically it's not about changing `isSelected` but rerendering `InspectorControls` component on mobile, when there is an update in component which contains controls (block edit). Unfortunately sometimes even rendering the whole `FlatList` makes updates in each block component :|

You should definitely do whatever is necessary to fix the issues with performance. I was mostly curious to learn if there is something fundamentally wrong with how inspector and block controls work. We didn't encounter similar issues on the web. Maybe those fills are doing more than checking the value in the block edit context and returning early with nothing rendered when `isSelected` is `false` 🤷🏻

**Review by dcalhoun — COMMENTED** [review]
@lukewalczak I had similar thoughts and questions as @gziolo in regards to reusing [the existing `isSelected` check](https://github.com/WordPress/gutenberg/blob/5f525d4d68574db2e0fec0de891b4bfa875b1808/packages/block-editor/src/components/inspector-controls/index.native.js#L22-L24) found in both web and native. Namely — if we already have this `isSelected` check one level down within `InspectorControls` why add this new check?

Is it accurate to say that the answer to that question is to hoist the conditional rendering higher up in the tree to (1) render (and re-render) fewer components and (2) reuse `isSelected` that is already available in the higher component? I.e. we already have `isSelected` in the parent (`edit.js`) and by placing the conditional render in the parent we (1) avoid looking up the `isSelected` again and (2) reduce the depth of the tree by avoiding `InspectorControls` altogether. Is that accurate?

A follow up question I have is if we are adding `isSelected` checks to the parents, should we remove the preexisting check from within `InspectorControls`? The argument I see _for removing it_ would be an attempt to avoid promoting new code relying upon the check with lower performance. The argument _against_ would likely be avoiding breaking existing usages.

Thanks for putting this PR together. I look forward to hearing your thoughts.

**Comment by lukewalczak** [comment]
Hello @dcalhoun, thanks for your comment! Let me refer to your concerns: 

> Is it accurate to say that the answer to that question is to hoist the conditional rendering higher up in the tree to (1) render (and re-render) fewer components and (2) reuse isSelected that is already available in the higher component? I.e. we already have isSelected in the parent (edit.js) and by placing the conditional render in the parent we (1) avoid looking up the isSelected again and (2) reduce the depth of the tree by avoiding InspectorControls altogether. Is that accurate?

Correct! Using the condition in the parent we can confirm whether component is selected and avoid re-rendering `InspectorControl` when there are updates in component.

> A follow up question I have is if we are adding isSelected checks to the parents, should we remove the preexisting check from within InspectorControls? The argument I see for removing it would be an attempt to avoid promoting new code relying upon the check with lower performance. The argument against would likely be avoiding breaking existing usages.

Definitely, I would like to avoid deletion the preexisting check from `InspectorControls`. Also I'm aware that adding the check before using `InspectorControls` in the _parent_ is not the best way, because actually we are duplicating the check and in the future we can easily forget about adding it.

I double checked the performance on the big post and results differences are are not as significant as I assumed before, hence I think we can close that PR.

File: packages/components/src/mobile/cycle-select-control/index.native.js [hunk 1]
```diff
@@ -7,33 +7,35 @@ import { memo } from '@wordpress/element';
...
-	label,
-	multiple = false,
-	onChange,
-	options = [],
-	className,
-	hideLabelFromVision,
-	...props
-} ) {
-	const id = `inspector-select-control-${ instanceId }`;
+const CycleSelectControl = memo(
```
[the remark this answers is not in the document]
Comment (lukewalczak) [reply packages/components/src/mobile/cycle-select-control/index.native.js:10]: In both comments I think I messed during rebasing.

**Comment by dcalhoun** [comment]
@lukewalczak ah, OK. Closing is fine if you believe that to be the best outcome. Thank you for helping me understand the entire context.

**Pull request closed without merging**